src/transforms/kubernetes/mod.rs

```diff
@@ -103,108 +143,103 @@ enum BuildError {
-        add_fields.sort();
-        add_fields.dedup();
-
-        let mut fields = Vec::new();
-        let mut unknown = Vec::new();
-        let mut all_fields = all_fields();
-
-        for field in add_fields {
-            if let Some(fun) = all_fields.remove(field.as_str()) {
-                fields.push(fun);
-            } else {
-                unknown.push(field.to_owned());
-            }
-        }
-
-        if unknown.is_empty() {
+        if config.max_retry_timeout > 0 {
             Ok(Self {
-                fields,
+                fields: config.fields.clone(),
                 metadata,
                 client,
                 delete_queue: VecDeque::new(),
+                cache_ttl: Duration::from_secs(config.cache_ttl),
+                max_retry_timeout: Duration::from_secs(config.max_retry_timeout),
             })
         } else {
-            Err(BuildError::UnknownFields { fields: unknown })
+            Err(BuildError::TooSmallMaxRetryTimeout)
         }
     }
 
     /// Listens for pod metadata changes and updates metadata map.
-    async fn run(mut self) -> BuildError {
+    async fn run(mut self) -> Result<(), BuildError> {
         let mut version = None;
         let mut error = None;
+        // Since this transform will in most cases be deployed on all Nodes
+        // and fairly simulationisly, we can immediately start with user
+        // defined max_retry_timeout.
+        let mut retry_timeout = self.max_retry_timeout;
         loop {
+            // Wait for bit before trying to watch again.
+            // We are sampling from a uniform distribution here.
+            let timeout = rand::thread_rng().gen_range(Duration::default(), retry_timeout);
+            let _ = Delay::new(Instant::now() + timeout)
+                .compat()
+                .await
+                .expect("Timer not set.");
+
             // Build watcher stream
-            let mut watcher = match self.client.watch_metadata(version.clone(), error.take()) {
-                Ok(watcher) => watcher,
-                Err(error) => return BuildError::WatchStreamBuild { source: error },
-            };
+            let mut watcher = self
+                .client
+                .watch_metadata(version.clone(), error.take())
+                .context(WatchStreamBuild)?
+                .compat();
             info!("Watching Pod metadata.");
 
             // Watch loop
-            let runtime_error = loop {
-                break match watcher.into_future().compat().await {
-                    Ok((Some(event), tail)) => {
-                        watcher = tail;
+            error = Some(RuntimeError::WatchUnexpectedlyEnded);
+            retry_timeout = self.max_retry_timeout.min(retry_timeout * 2);
 
+            while let Some(next) = watcher.next().await {
+                match next {
+                    Ok(event) => {
                         self.delete_update();
```

Comment:
What happens if we get no new updates, the only way for the cache to get invalidated is via a new event which then triggers cache garbage collection, right? Would it make sense to select on a timer here?

Reply:
Yes, that is the only way. And since that's also the only way for the cache to get bigger, using a timer won't make any more difference than freeing some kilobytes sooner rather than later.

Reply:
Well a timer would allow us to avoid issues with failing to connect to kube right? I am not sure I see a disadvantage to using a timer we do this in a few other places like batching.

Reply:
What issues do you mean? Regarding postponing/never deleting entries of deleted Pod's, I don't see any way of them becoming a problem regardless of any issue.

Reply:
I'll add it, then we can decide if the added complexity is worth it.